### PR TITLE
#[UIC-1324]:Enabled knox sso using hadoop jwt token parsing

### PR DESCRIPTION
**Pull request by jitendra-kumawat**
### CATEGORY

Choose one

- [ ] Bug Fix
- [x] Enhancement (new features, refinement)
- [ ] Refactor
- [ ] Add tests
- [ ] Build / Development Environment
- [ ] Documentation

### SUMMARY
Enabled knox sso using hadoop jwt token parsing

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### REVIEWERS
@ankursinghal2005

### CODE REVIEW

File: docker/superset_config.py [hunk 1]
```diff
@@ -122,3 +122,8 @@ class CeleryConfig(object):
 
 
 CELERY_CONFIG = CeleryConfig
+
+# KnoxSSO-superset integration constants
+KNOX_SSO_ENABLED = boolify(get_env_variable('KNOX_SSO_ENABLED'))
+KNOX_PUBLIC_KEY = get_env_variable('KNOX_PUBLIC_KEY')
```
Comment (arpit-agarwal) [inline docker/superset_config.py]: Should we rename this to `KNOX_SSO_PUBLIC_KEY`

File: docker/superset_config.py [hunk 2]
```diff
@@ -122,3 +122,8 @@ class CeleryConfig(object):
 
 
 CELERY_CONFIG = CeleryConfig
+
+# KnoxSSO-superset integration constants
+KNOX_SSO_ENABLED = boolify(get_env_variable('KNOX_SSO_ENABLED'))
```
Comment (arpit-agarwal) [inline docker/superset_config.py]: Should we rename this to IS_KNOX_SSO_ENABLED

File: superset/config.py [hunk 3]
```diff
@@ -586,6 +586,8 @@ class CeleryConfig(object):
 IS_EPOCH_S_TRULY_UTC = False
 
 
+# set on KNOX_SSO_ENABLED here ,because of unit test cases and rest KNOX varibales are defined in superset_config.py 
+KNOX_SSO_ENABLED = False
```
Comment (arpit-agarwal) [inline superset/config.py]: should we rename this to IS_KONX_SSO_ENABLED

File: superset/sso_auth.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+def get_token_contents(token):
+    contents = jwt.decode(token,get_publickey())
+    return contents
+
+def _find_user(username, sm):
+    """extracted from flask_appbuilder.security.manager.BaseSecurityManager.find_user"""
+    user = sm.find_user(username)
+    if not user:
+        user = security_manager.add_user(
```
Comment (arpit-agarwal) [inline superset/sso_auth.py]: should we use `sm.add_user`. what is the difference between `sm` and `security_manager ` ?

File: superset/sso_auth.py [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+                last_name=username,
+                email=username + '@email.notfound',
+                role=security_manager.find_role("Admin")
+            )
+    return user
+
+def parse_hadoop_jwt():
+    if(config.KNOX_SSO_ENABLED):
+        logging.info("Attaching JWT handler")
+        jwt_token = request.cookies.get("hadoop-jwt", None)
```
Comment (arpit-agarwal) [inline superset/sso_auth.py]: This name of token can be configurable. Can we take this from config and default to `hadoop-jwt` if not provided

File: superset/sso_auth.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+        logging.debug("Token: %s"%jwt_token)
+        
+        sso_login_url = config.KNOX_SSO_URL+"?originalUrl="+request.url
```
Comment (arpit-agarwal) [inline superset/sso_auth.py]: `originalUrl ` is a token that can be modified in KNOX configuration. please expose it as a variable and default to `originalUrl`

File: superset/sso_auth.py [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+            token_contents = get_token_contents(jwt_token) 
+        except jwt.ExpiredSignatureError:
+                logging.error("Signature expired. Please log in again.")  
+                return redirect(sso_login_url)
+        except jwt.InvalidTokenError:
+                logging.error("Invalid token. Please log in again.")  
+                return redirect(sso_login_url)
+
+        token_expiry = int(token_contents['exp'])
+        now = int(datetime.now(tz=timezone.utc).timestamp())
```
Comment (arpit-agarwal) [inline superset/sso_auth.py]: I am not sure but if FAB has a variable to specify application timezone we should use that here

Comment (jitendra-kumawat) [reply superset/sso_auth.py]: No variable at app level but changed it to local tz like tz.tzlocal()

**Pull request merged**